src/signature.rs

```diff
@@ -0,0 +1,167 @@
+use std::borrow::Borrow;
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+use std::ops::{Add, AddAssign, Div, Mul, MulAssign, Neg, Shl, Sub, SubAssign};
+use std::str::FromStr;
+
+use num::bigint::ParseBigIntError;
+use num::BigUint;
+use num::pow;
+
+use crate::{Expression, GadgetBuilder, BooleanExpression, EdwardsCurve, Curve, CurvePoint, EdwardsPointExpression, CompressionFunction};
+use crate::field::{Element, Field};
+
+pub trait SignatureExpression<F: Field, C: EdwardsCurve<F>, CF> {
+    fn verify(
+        &self,
+        builder: &mut GadgetBuilder<F>,
+        message: &Expression<F>,
+        public_key: &EdwardsPointExpression<F, C>,
+        compress: &CF,
+    ) where CF: CompressionFunction<F>;
+}
+
+/// Struct to represent a Schnorr Signature.
+///
+/// Assumes that the message has already been hashed to a field element
+/// Signature is a tuple consisting of scalars (s, e), where r_v = sg + ey
+/// Public key is a group element, y = xg for private key x
+pub struct SchnorrSignatureExpression<F: Field> {
+    s: Expression<F>,
+    e: Expression<F>,
+}
+
+impl<F: Field, C: EdwardsCurve<F>, CF> SignatureExpression<F, C, CF> for SchnorrSignatureExpression<F> {
+
+    /// signature.verify(builder, message, public_key, compression_function)
```

Comment:
I think the docs (https://docs.rs/r1cs/) will look nicer if we omit this, since similar info is included automatically